src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,972 @@
+                editor.InsertParameter(containingMethod, insertionIndex, parameter);
+            }
+
+            return editor.GetChangedRoot();
+        }
+
+        private static async Task<ITypeSymbol> GetTypeOfExpressionAsync(Document document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var typeSymbol = semanticModel.GetTypeInfo(expression, cancellationToken).ConvertedType ?? semanticModel.Compilation.ObjectType;
+            return typeSymbol;
+        }
+        /// <summary>
+        /// This method iterates through the variables in the expression and maps the variables back to the parameter
+        /// it is associated with. It then maps the parameter back to the argument at the invocation site and gets the
+        /// index to retrieve the updated arguments at the invocation.
+        /// </summary>
+        private TExpressionSyntax CreateNewArgumentExpression(SyntaxEditor editor, ISyntaxFactsService syntaxFacts,
+            Dictionary<TIdentifierNameSyntax, IParameterSymbol> mappingDictionary,
+            ImmutableDictionary<IParameterSymbol, int> parameterToArgumentMap,
+            SeparatedSyntaxList<SyntaxNode> updatedInvocationArguments)
+        {
+            foreach (var (variable, mappedParameter) in mappingDictionary)
+            {
+                var parameterMapped = parameterToArgumentMap.TryGetValue(mappedParameter, out var index);
+                if (parameterMapped)
+                {
+                    var updatedInvocationArgument = updatedInvocationArguments[index];
+                    var argumentExpression = syntaxFacts.GetExpressionOfArgument(updatedInvocationArgument);
+                    var parenthesizedArgumentExpression = editor.Generator.AddParentheses(argumentExpression, includeElasticTrivia: false);
+                    editor.ReplaceNode(variable, parenthesizedArgumentExpression);
+                }
+                else if (mappedParameter.HasExplicitDefaultValue)
+                {
+                    var generatedExpression = GenerateExpressionFromOptionalParameter(mappedParameter);
+                    var parenthesizedGeneratedExpression = editor.Generator.AddParentheses(generatedExpression, includeElasticTrivia: false);
+                    editor.ReplaceNode(variable, parenthesizedGeneratedExpression);
+                }
+            }
+
+            return (TExpressionSyntax)editor.GetChangedRoot();
+        }
+
+        private static ImmutableDictionary<IParameterSymbol, int> MapParameterToArgumentsAtInvocation(
+            ISemanticFactsService semanticFacts, SeparatedSyntaxList<SyntaxNode> arguments,
+            SemanticModel invocationSemanticModel, CancellationToken cancellationToken)
+        {
+            var mapping = ImmutableDictionary.CreateBuilder<IParameterSymbol, int>();
```

Comment:
so we're going to create a *lot* of these temp dictionaries.  a nicer way woudl be to have one 'scratch dictionary' we careate at teh start of the loop that calls this, which passes it in, asking it to be filled.  the caller clears the dictionary before calling into this.

Reply:
fixed